src/AISmart.GAgent.Core/GAgentBase.cs

```diff
@@ -191,91 +202,47 @@ protected async Task<Guid> PublishAsync<T>(T @event) where T : EventBase
 
     private async Task PublishAsync<T>(EventWrapper<T> eventWrapper) where T : EventBase
     {
-        if (_publishers.Count == 0)
+        await LoadPublishersAsync();
+        if (_publishers.State.Count == 0)
         {
             return;
         }
 
-        foreach (var publisher in _publishers.Select(kp => kp.Value))
+        foreach (var publisher in _publishers.State.Select(kp => kp.Value))
         {
-            await publisher.OnNextAsync(eventWrapper);
+            var stream = GetStream(publisher);
+            await stream.OnNextAsync(eventWrapper);
         }
     }
 
-    private async Task AckAsync(EventWrapper<TEvent> eventWrapper)
+    private async Task SubscribeAsync(IAsyncStream<EventWrapperBase> stream)
     {
-        // var pubAgent = _clusterClient.GetGrain<IAgent<TState>>(eventWrapper.GrainId);
-        //
-        // try
-        // {
-        //     await ExecuteAsync(eventWrapper.Event);
-        // }
-        // catch (Exception e)
-        // {
-        //     Logger.LogError(e, "Error executing event");
-        // }
-        //
-        // await ((pubAgent as GAgent<TState,TEvent>)!).DoAckAsync(eventWrapper);
+        var streamId = stream.StreamId;
+        foreach (var observer in Observers.Keys)
+        {
+            var handle = await stream.SubscribeAsync(observer);
+            var handleId = handle.HandleId;
+            Observers[observer][streamId] = handleId;
+        }
     }
 
-    private async Task DoAckAsync(EventWrapper<TEvent> eventWrapper)
+    public override async Task OnActivateAsync(CancellationToken cancellationToken)
```

Comment:
Since we are adding important logic in OnActivateAsync, maybe we should seal it and introduce our own OnActivateAsync for the user to override in case he did not call base.OnActivateAsync